src/vs/editor/contrib/codeAction/browser/codeActionMenu.ts

```diff
@@ -305,28 +366,107 @@ export class CodeActionMenu extends Disposable implements IEditorContribution {
 		renderDisposables.add(this.codeActionList.value.onDidChangeSelection(e => this._onListSelection(e)));
 		renderDisposables.add(this._editor.onDidLayoutChange(e => this.hideCodeActionWidget()));
 
-		// Populating the list widget and tracking enabled options.
+		enum CodeActionArrayGroup {
+			quickfix = 0,
+			extract = 1,
+			convert = 2,
+			normalRefactor = 3,
+			separator = 4,
+			other = 5,
+		}
+
+		// Filters and groups code actions by their group
+		const menuEntries: IAction[][] = [[], [], [], [], [], []];
```

Comment:
I think this data structure will be difficult to maintain and scale. What if you instead created arrays/data structures for each group:

```
const quickfixGroup = [];
...
```